### PR TITLE
Replace bespoke x25519 with a library

### PR DESCRIPTION
**Pull request by jeff-hiner**
This completely removes the bespoke X25519 implementation, replacing it with the `x25519_dalek` crate.

I couldn't use `ring` because it doesn't allow DH agreements with static keys, nor does it allow "reusable" ephemeral keys required for Noise implementations. Turns out `x25519_dalek` is far more ergonomic anyway.

This change removes approximately 1000 lines of code, and on my machine improves the x25519 key generation speed by a factor of 3.

### CODE REVIEW

File: boringtun/Cargo.toml [hunk 1]
```diff
@@ -17,8 +17,9 @@ parking_lot = "0.12"
 tracing = "0.1.29"
 ip_network = "0.4.1"
 ip_network_table = "0.2.0"
-zeroize = { version = "1.5.2", features = ["zeroize_derive"]}
 ring = "0.16"
+x25519-dalek = { version = "1.2", features = ["reusable_secrets"] }
+rand_core = "0.5.1"
```
Comment (jeff-hiner) [inline boringtun/Cargo.toml:22]: This is an older version of `rand_core` until x25519_dalek 2.0 lands.

File: boringtun/src/device/integration_tests/mod.rs [hunk 2]
```diff
@@ -108,8 +110,8 @@ mod tests {
             // The local endpoint port is the remote listen port
             conf.push_str(&format!("\nListenPort = {}", self.endpoint.port()));
             conf.push_str(&format!(
-                "\nPrivateKey = {}",
```
Comment (Noah-Kennedy) [inline boringtun/src/device/integration_tests/mod.rs:111]: What does this change do?

Comment (Noah-Kennedy) [inline boringtun/src/device/integration_tests/mod.rs:111]: What does this change do?

File: boringtun/src/noise/handshake.rs [hunk 3]
```diff
@@ -30,25 +32,22 @@ static INITIAL_CHAIN_HASH: [u8; KEY_LEN] = [
 
 macro_rules! HASH {
     ($data1:expr, $data2:expr) => {
-        Blake2s::new_hash().hash(&$data1).hash(&$data2).finalize()
```
Comment (Noah-Kennedy) [inline boringtun/src/noise/handshake.rs:33]: Can we get rid of these macros? We should not be mixing macros and cryptography.

File: boringtun/src/device/api.rs [hunk 4]
```diff
@@ -155,7 +156,7 @@ impl Device {
 fn api_get(writer: &mut BufWriter<&UnixStream>, d: &Device) -> i32 {
     // get command requires an empty line, but there is no reason to be religious about it
     if let Some(ref k) = d.key_pair {
-        writeln!(writer, "private_key={}", encode_hex(k.0.as_bytes()));
```
Comment (Noah-Kennedy) [inline boringtun/src/device/api.rs:158]: What are we changing here?

File: boringtun/src/device/integration_tests/mod.rs (hunk 2 above)
Comment (jeff-hiner) [reply boringtun/src/device/integration_tests/mod.rs:111]: Secrets can't be trivially exported with x25519_dalek. You can _move_ the secret out, and you can run Diffie-Hellman operations with it, but you can't just get a ref and display it raw.

File: boringtun/src/noise/handshake.rs (hunk 3 above)
Comment (jeff-hiner) [reply boringtun/src/noise/handshake.rs:33]: Not easily, no, and certainly not in this PR. These are in place of the KDF lines in the specification to make it easier to read and audit. Surprisingly, they do actually make sense here. The problem is the use of `&` inside the macro, which does harm readability.

File: boringtun/src/device/api.rs (hunk 4 above)
Comment (jeff-hiner) [reply boringtun/src/device/api.rs:158]: Again, dumping a private key like this is a terrible idea.

**Review by Noah-Kennedy — APPROVED** [review]
Looks good, let's merge.

**Comment by Noah-Kennedy** [comment]
@jeff-hiner we've got conflicts sadly. Also, this integration test can be deleted. I'm about to redo the tests anyways.

File: boringtun/src/device/integration_tests/mod.rs (hunk 2 above)
Comment (jeff-hiner) [reply boringtun/src/device/integration_tests/mod.rs:111]: Looks like this is what's breaking the integration tests. `gen_wg_conf` is trying to generate a config file for a fake server. That server needs a private key.

**Pull request merged**